Add support for multiple backends

robin-nitrokey opened the pull request:
This is a reduced version of https://github.com/trussed-dev/trussed/pull/41 as discussed today.  The main difference is that the service endpoints cannot be configured at runtime (using a syscall), but only during client creation.  This makes the interface much simpler.  I did not convert the backend ID to an integer as originally planned because there is only one remaining struct that uses the backend ID, `ServiceEndpoint`.

Based on https://github.com/trussed-dev/trussed/pull/67

robin-nitrokey commented:
This backends implementation only changes the public interface for a single function, `Service::add_endpoint`, which is probably not used outside Trussed anyway.  If this is a concern, we can introduce a variant to maintain compatibility.

robin-nitrokey commented:
Rebased onto 88de4c6b3ac8c94534aad98efbed02af8eaec70c.

nickray reviewed (changes requested):
This can be reverted, https://github.com/solokeys/solo2/pull/152 is merged. Thanks for that!

nickray reviewed (changes requested):
I explored the suggestions concretely here: https://github.com/robin-nitrokey/trussed/compare/backends...nickray:trussed:elaboration

If this seems OK to you, happy to push to your branch and merge this PR.

robin-nitrokey commented:
Thanks for the review!  While working on https://github.com/trussed-dev/trussed/pull/70 I realized that the current `Backends::select` method is not sufficient:  We have to add an associated type for the context to `Backend` which makes the trait non-object-safe, so we cannot return a reference to it.  Instead, we directly have to call the `Backend::reply_to` method for a given backend:

```rust
pub trait Backends<P: Platform> {
    type Id: 'static;
    type ExtensionId: TryFrom<u8, Error = Error>;
    type Context: Default;

    fn core_request(
        &mut self,
        id: &Self::Id,
        ctx: &mut Context<Self::Context>,
        request: &Request,
        resources: &mut ServiceResources<P>,
    ) -> Result<Reply, Error> {
        let _ = (id, ctx, request, resources);
        Err(Error::RequestNotAvailable)
    }

    fn extension_request(
        &mut self,
        backend: &Self::Id,
        extension: &Self::ExtensionId,
        ctx: &mut Context<Self::Context>,
        request: &request::Extension,
        resources: &mut ServiceResources<P>,
    ) -> Result<reply::Extension, Error> {
        let _ = (backend, extension, ctx, request, resources);
        Err(Error::RequestNotAvailable)
    }
}
```

So I would rather keep the `Backends` name or maybe use something more generic like `Dispatch`.

nickray commented:
Dispatch sounds good. I find Backends too close to Backend. From the sound of it, though this PR needs to simmer a bit until #70 is more final?

robin-nitrokey commented:
Okay, I’ll change it to `Dispatch`.  #70 would already be ready to review if this is merged.  I split it up into two PRs to make it easier to review, but if you’d prefer doing both in one PR, we can also do that.

robin-nitrokey commented:
I’ve updated the PR:
- Rebased onto 1f3cb9db10c21a57f871725b339677c936a61b47
- Renamed `Backends` to `Dispatch` per our discussion
- Added a commit that introduces backend-specific context (extracted from #70)

sosthene-nitrokey commented on src/client.rs:
```diff
@@ -685,6 +687,70 @@ pub trait UiClient: PollClient {
+}
+
+impl<B: 'static> ClientBuilder<B> {
+    pub fn with_backends<C: 'static>(self, backends: &'static [BackendId<C>]) -> ClientBuilder<C> {
```
I would remove the `with`, it is generally not used in the builder pattern.

robin-nitrokey commented:
Updated:
- Rebased onto 47afe55313ef418c9448b690c209a3c4f1c8fd64.
- Renamed `ClientContext` to `CoreContext`.
- Renamed `with_backends` in `ClientBuilder` to `backends`.

nickray commented on src/backend.rs:
```diff
@@ -0,0 +1,48 @@
+    }
+}
+
+impl<P: Platform> Dispatch<P> for () {
```
I would prefer having this implemented on a specific type `struct SoftwareOnly` instead of using the unit type (name to be bikeshed). Think this is more clear.

nickray commented on src/backend.rs:
```diff
@@ -0,0 +1,48 @@
+}
+
+impl<P: Platform> Dispatch<P> for () {
+    type BackendId = ();
```
I would prefer using a void type `enum Empty {}` here instead of using the unit type (name to be bikeshed).
Cf. https://github.com/robin-nitrokey/trussed/compare/backends...nickray:trussed:elaboration even if the Dispatch method changed now. This enforces that you can't even generate a BackendId::Custom.

robin-nitrokey replied:
Done.

robin-nitrokey replied:
Could probably be `!` once it’s stable.

nickray commented on src/backend.rs:
```diff
@@ -0,0 +1,48 @@
+
+impl<P: Platform> Dispatch<P> for () {
+    type BackendId = ();
+    type Context = ();
```
Not sure what a "more correct" default type (empty struct vs empty enum) and name would be here.

robin-nitrokey replied:
I think the empty tuple is fine here because it just means “no data”.

nickray replied:
So `NoData`? :)

nickray commented on src/client.rs:
```diff
@@ -685,6 +687,69 @@ pub trait UiClient: PollClient {
+        self,
+        service: &mut Service<P, D>,
+        syscall: S,
+    ) -> Result<ClientImplementation<S>, ()> {
```
In these Results with trivial error types, maybe we should use Option and mark with `#[must_use]`?
I'm finding the proliferation of `()` a bit hard to follow.
Or would this break `?` in certain places?

robin-nitrokey replied:
I just used `()` for consistency with the existing methods, but I agree that `Error` is much better.  I’ll add a `ClientCountExceeded` variant and change all relevant methods in `ClientBuilder` and `Service` to return that error instead.

nickray replied:
Yeah that works. Generally it'd be better to have tight error types (i.e. a client of just this method would know which error variants it can actually expect), but doing this in Rust is tedious as you'd need tons of conversion traits.

robin-nitrokey commented:
Updated:
- Added `Error::ClientCountExceeded` variant and used it in the relevant methods.
- Added `SoftwareOnly` struct as the baseline `Dispatch` implementation.
- Added `Empty` enum.

I’m not sure whether “Software” is the best term to describe the Trussed default backend.  I think it was chosen with backends like SE050 in mind, but we will probably also have software-only backends that just implement extensions.  Maybe “Core” would be better?

nickray commented:
I suggested `SoftwareOnly` for consistency with `BackendId::Software` (so change both if you change one). `Core` would be consistent with your `CoreContext`. `Default` or `Baseline` or whatever alternatives... Agree that the configured default backend doesn't necessarily have to be software, nor a custom backend non-software.

nickray commented on src/backend.rs:
```diff
@@ -0,0 +1,43 @@
+    types::{ClientContext, Empty},
+};
+
+pub enum BackendId<C> {
```
Can we have `I` instead of `C` here please?

nickray commented on src/client.rs:
```diff
@@ -685,6 +687,65 @@ pub trait UiClient: PollClient {
     }
 }
 
+pub struct ClientBuilder<B: 'static = Empty> {
```
`I` instead of `B`. It's a bit tricky to stay on top of Backend vs Id throught this code, I think this would help.

nickray commented on src/client.rs:
```diff
@@ -685,6 +687,65 @@ pub trait UiClient: PollClient {
+}
+
+impl<B: 'static> ClientBuilder<B> {
+    pub fn backends<C: 'static>(self, backends: &'static [BackendId<C>]) -> ClientBuilder<C> {
```
Perhaps `J` instead of `C`?

nickray commented on src/client.rs:
```diff
@@ -685,6 +687,65 @@ pub trait UiClient: PollClient {
+    }
+}
+
+impl<B: 'static> ClientBuilder<B> {
```
`I` -> `B`

nickray commented on src/pipe.rs:
```diff
@@ -71,11 +72,12 @@ cfg_if::cfg_if! {
 // https://doc.micrium.com/display/osiiidoc/Using+Message+Queues
 
-pub struct ServiceEndpoint {
+pub struct ServiceEndpoint<B: 'static = ()> {
```
`I`

robin-nitrokey commented:
Updated:
- Changed `B` to `I` for the backend ID (and `C` to `J`).
- Added `NoData` struct (though I’m not really convinced ;)).
- Added documentation.
- Changed “software” to “core” when referring to the default request implementation in Trussed.

robin-nitrokey commented:
Rebased onto 5c9bc60eaecf7d38b3512b973ac74309980ae721 and added to changelog.

nickray commented on src/types.rs:
```diff
@@ -233,17 +241,36 @@ pub mod consent {
 // subtree in the filesystem (see docs in src/store.rs) and to maintain
 // the walker state of the directory traversal syscalls.
-pub struct ClientContext {
+pub struct CoreContext {
```
Shouldn't this be non-exhaustive in light of other discussions?

robin-nitrokey replied:
I’ll add it to #71.  It has to be part of the public API so that the backends have access to it.

daringer commented:
:+1:
